qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py

```diff
@@ -223,22 +134,205 @@ def mcrz(
     """
     control_qubits = self.qbit_argument_conversion(q_controls)
     target_qubit = self.qbit_argument_conversion(q_target)
+
     if len(target_qubit) != 1:
         raise QiskitError("The mcrz gate needs a single qubit as target.")
     all_qubits = control_qubits + target_qubit
-    target_qubit = target_qubit[0]
     self._check_dups(all_qubits)
 
     n_c = len(control_qubits)
-    if n_c == 1:  # cu
-        _apply_cu(self, 0, 0, lam, control_qubits[0], target_qubit, use_basis_gates=use_basis_gates)
+    if n_c <= 6:
+        ncrz = ControlRotationGate(lam, n_c, axis="z")
     else:
-        lam_step = lam * (1 / (2 ** (n_c - 1)))
-        _apply_mcu_graycode(
-            self, 0, 0, lam_step, control_qubits, target_qubit, use_basis_gates=use_basis_gates
-        )
+        rzgate = numpy.cos(lam / 2) * numpy.eye(2) - 1j * numpy.sin(lam / 2) * ZGate().__array__()
+        ncrz = MCU2Gate(rzgate, n_c, label=f"Rz({lam:0.3f})")
+    # if use_basis_gates:
+    #     ncrz = transpile(ncrz, basis_gates=['cx','u', 'p'])
+
+    self.append(ncrz, [*control_qubits, q_target])
 
 
 QuantumCircuit.mcrx = mcrx
 QuantumCircuit.mcry = mcry
 QuantumCircuit.mcrz = mcrz
+
+
+class ControlRotationGate(ControlledGate):
```

Comment:
Hi @AlexisRalli, I think that ucry, ucrx and ucrz in this [folder](https://github.com/Qiskit/qiskit-terra/tree/main/qiskit/extensions/quantum_initializer) already implements Theorem 8 of https://arxiv.org/pdf/quant-ph/0406176.pdf.